trieste/models/keras/architectures.py

```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+        :param output_tensor_spec: Tensor specification for the output of the network.
+        :param hidden_layer_args: Specification for building dense hidden layers. Each element in
+            the sequence should be a dictionary containing arguments (keys) and their values for a
+            :class:`~tf.keras.layers.Dense` hidden layer. Please check Keras Dense layer API for
+            available arguments. Objects in the sequence will sequentially be used to add
+            :class:`~tf.keras.layers.Dense` layers. Length of this sequence determines the number of
+            hidden layers in the network. Default value is two hidden layers, 50 nodes each, with
+            ReLu activation functions. Empty sequence needs to be passed to have no hidden layers.
+        :param rate: Probability of dropout assigned to each layer of a fully connected network.
+            If scalar, the same probability will be assigned to each layer. Otherwise accepts a list of 
+            probabilities corresponding to each layer. 
+        :raise ValueError: If objects in ``hidden_layer_args`` are not dictionaries.
+        """
+        super().__init__(input_tensor_spec, output_tensor_spec)
+
+        self._hidden_layer_args = hidden_layer_args
+        self.rate = rate
+        self._model = self._build_model()
+
+    @property
+    def model(self) -> tf.keras.Model:
+        return self._model
+
+    @property
+    def rate(self) -> Sequence[float]:
+        return self._rate
+    
+    @rate.setter
+    def rate(self, rate):
+        if isinstance(rate, Sequence):
+            if not len(rate) == len(self._hidden_layer_args) + 1:
+                raise ValueError(
+                    f"""Requires a dropout probability for each hidden layer and the output layer. 
+                    Got {len(rate)} dropout probabilities for {len(self._hidden_layer_args) + 1} layers."""
+                )
+            for p in rate:
+                self._check_probability(p)
+            self._rate = rate
+        elif isinstance(rate, float):
+            self._check_probability(rate)
+            self._rate = [rate for _ in range(len(self._hidden_layer_args) + 1)]
+        else:
+            raise TypeError(f"dropout_prob needs to be a sequence, float or int. Instead got {type(rate)}")
+
+    def _check_probability(self, p: float) -> None:
+        if not 0 < p < 1:
+            raise ValueError(
+                f"Invalid probability {p} received."
+            )
+
+    def _gen_input_tensor(self) -> tf.keras.Input:
+
+        input_tensor = tf.keras.Input(
+            shape=self.input_tensor_spec.shape,
+            dtype=self.input_tensor_spec.dtype,
+            name=self.input_layer_name
+        )
+        return input_tensor
+
+    def _gen_hidden_layers(self, input_tensor: tf.Tensor) -> tf.Tensor:
+
+        for index, hidden_layer_args in enumerate(self._hidden_layer_args):
+            layer_name = f"{self.network_name}dense_{index}"
```

Comment:
i needed to set names carefully bcs dealing with ensemble of models is a bit more involved (in terms of assigning inputs/outputs correctly), so you can dispense with the custom naming